cmd/ignition2rpm/main.go

```diff
@@ -0,0 +1,404 @@
+		packedRPM.RPMMetaData.Provides = append(packedRPM.RPMMetaData.Provides, &rpmpack.Relation{Name: "rpmostree(override)"})
+	}
+
+	if err != nil {
+		glog.Fatalf("Failed to create new RPM file: %s", err)
+	}
+
+	/*Adding prefixes alone is not enough to make it relocatable (nope, doesn't work)
+	magic numbers come from rpmtag.h
+	packedRPM.AddCustomTag(1098, rpmpack.EntryStringSlice([]string{"/etc/", "/var/", "usr"}))
+	*/
+
+	// have to process the envelope if it's machineconfig, otherwise we don't
+	switch c := configi.(type) {
+	case ign3types.Config:
+
+		//process the ignition into an RPM
+		err = Ign2Rpm(packedRPM, &c)
+		if err != nil {
+			glog.Fatalf("Unable to convert ignition to RPM payload: %v", err)
+		}
+	case mcfgv1.MachineConfig:
+
+		//shuck the ignition out of the machineconfig if need be
+		//TODO: fish the name out of the machineconfig for the metadata?
+		newIgnConfig, err := ctrlcommon.ParseAndConvertConfig(c.Spec.Config.Raw)
+		if err != nil {
+			glog.Fatalf("Unable to convert machine config to ignition: %v", err)
+		}
+		err = Ign2Rpm(packedRPM, &newIgnConfig)
+		if err != nil {
+			glog.Fatalf("Unable to convert ignition to RPM payload: %v", err)
+		}
+	}
+
+	//make the rpmfile we're going to write to disk
+	f, err := os.Create(outputRPM)
+	if err != nil {
+		panic(err)
+	}
+
+	//stuff our RPM file guts into it
+	if err := packedRPM.Write(f); err != nil {
+		glog.Fatalf("Unable to write RPM %s to disk: %v", outputRPM, err)
+	}
+
+	glog.Infof("Wrote %s", outputRPM)
+
+}
+
+type onceFromOrigin int
+
+const (
+	onceFromUnknownConfig onceFromOrigin = iota
+	onceFromLocalConfig
+	onceFromRemoteConfig
+)
+
+// blatantly stolen from mco daemon's once-from
+func senseAndLoadOnceFrom(onceFrom string) (interface{}, onceFromOrigin, error) {
+	var (
+		content     []byte
+		contentFrom onceFromOrigin
+	)
+	// Read the content from a remote endpoint if requested
+	/* #nosec */
+	if strings.HasPrefix(onceFrom, "http://") || strings.HasPrefix(onceFrom, "https://") {
+		contentFrom = onceFromRemoteConfig
+		resp, err := http.Get(onceFrom)
+		if err != nil {
+			return nil, contentFrom, err
+		}
+		defer resp.Body.Close()
+		// Read the body content from the request
+		content, err = ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return nil, contentFrom, err
+		}
+
+	} else {
+		// Otherwise read it from a local file
+		contentFrom = onceFromLocalConfig
+		absoluteOnceFrom, err := filepath.Abs(filepath.Clean(onceFrom))
+		if err != nil {
+			return nil, contentFrom, err
+		}
+		content, err = ioutil.ReadFile(absoluteOnceFrom)
+		if err != nil {
+			return nil, contentFrom, err
+		}
+	}
+
+	// Try each supported parser
+	ignConfig, err := ctrlcommon.ParseAndConvertConfig(content)
+	if err == nil && ignConfig.Ignition.Version != "" {
+		glog.V(2).Info("onceFrom file is of type Ignition")
+		return ignConfig, contentFrom, nil
+	}
+
+	// Try to parse as a machine config
+	mc, err := resourceread.ReadMachineConfigV1(content)
```

Comment:
I mentioned this elsewhere, but I'll also mention it here for posterity: This function will only return the first MachineConfig it finds in the byte array. Meaning that if you have multiple MachineConfigs within a single YAML file, only the first one will be read. The rest will be ignored.